Core/BookmarksStateValidation.swift

```diff
@@ -0,0 +1,91 @@
+//
+//  BookmarksStateValidation.swift
+//  DuckDuckGo
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import CoreData
+import Bookmarks
+import Persistence
+
+public class BookmarksStateValidation {
+
+    enum Constants {
+        static let bookmarksDBIsInitialized = "bookmarksDBIsInitialized"
+    }
+
+    public enum ValidationError {
+        case bookmarksStructureLost
+        case bookmarksStructureBroken
+        case validatorError
+    }
+
+    let keyValueStore: KeyValueStoring
+    let errorHandler: (ValidationError, Error?, [String: String]) -> Void
```

Comment:
@bwaresiak What do you think of the following change? Instead of having a callback with multiple parameters, we could leverage the `ValidationError` enum.

```swift
public enum ValidationError {
        case bookmarksStructureLost
        case bookmarksStructureBroken(additionalParams: [String: String])
        case validatorError(Error)
    }
```

If we do that it’s easier to know which validation error is returning what, also the code inside the `BookmarksDatabaseSetup` will look cleaner.

```swift
let validator = BookmarksStateValidation(keyValueStore: UserDefaults.app) { validationError in
    switch validationError {
    case .bookmarksStructureLost:
        DailyPixel.fire(pixel: .debugBookmarksStructureLost, includedParameters: [.appVersion])
    case .bookmarksStructureBroken(let additionalInfo):
        DailyPixel.fire(pixel: .debugBookmarksInvalidRoots,
                        withAdditionalParameters: additionalInfo,
                        includedParameters: [.appVersion])
    case .validatorError(let underlyingError):
        var params = [String: String]()
        let processedErrors = CoreDataErrorsParser.parse(error: underlyingError as NSError)
        params = processedErrors.errorPixelParameters

        DailyPixel.fireDailyAndCount(pixel: .debugBookmarksValidationFailed,
                                     withAdditionalParameters: params,
                                     includedParameters: [.appVersion])
    }
}
```

Reply:
Excellent idea :)